src/cache.rs

```diff
@@ -0,0 +1,394 @@
+        self
+    }
+
+    pub async fn build(self) -> ClientResult<SkytableCache<'a, ConnectionMgrTls, K, V>> {
+        let pool = skytable::pool::get_tls_async(
+            32,
+            skytable::Config::new(self.host, self.port, self.username, self.password),
+            self.certificate,
+        )
+        .await
+        .unwrap();
+
+        let cache = SkytableCache {
+            space: self.space,
+            model: self.model,
+            seconds: self.seconds,
+            pool,
+            refresh: self.refresh,
+            _phantom: PhantomData,
+        };
+        cache.init().await;
+
+        Ok(cache)
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum SkytableCacheError {
+    #[error(transparent)]
+    SkytableError(#[from] skytable::error::Error),
+    #[error(transparent)]
+    SerdeError(#[from] serde_json::Error),
+}
+
+#[derive(Clone, Query, Response)]
+struct CachedSkytableValue {
+    pub ckey: String,
+    pub cvalue: String,
+    pub ttl: i64,
+}
+impl CachedSkytableValue {
+    fn new(ckey: String, cvalue: String, ttl: i64) -> Self {
+        Self { ckey, cvalue, ttl }
+    }
+}
+
+pub struct SkytableCache<'a, C: ManageConnection, K, V> {
+    space: &'a str,
+    model: &'a str,
+    seconds: u64,
+    pool: bb8::Pool<C>,
+    refresh: bool,
+    _phantom: PhantomData<(K, V)>,
+}
+
+impl<'a, C, I, S, K, V> SkytableCache<'a, C, K, V>
+where
+    S: AsyncRead + AsyncWrite + Send + Sync + Unpin,
+    I: DerefMut<Target = TcpConnection<S>> + Send + Sync,
+    C: ManageConnection<Connection = I> + Send + Sync,
+{
+    fn build_model(&self) -> String {
+        format!("{}.{}", self.space, self.model)
+    }
+
+    pub async fn init(&self) {
+        let mut connection = self.pool.get().await.unwrap();
+
+        let pipeline = Pipeline::new()
+            .add(&query!(format!(
+                "CREATE SPACE IF NOT EXISTS {}",
+                self.space
+            )))
+            .add(&query!(format!(
+                "CREATE MODEL IF NOT EXISTS {}.{}(ckey: string, cvalue: string, ttl: sint64)",
+                self.space, self.model
+            )));
+        connection.execute_pipeline(&pipeline).await.unwrap();
+    }
+}
+
+#[async_trait::async_trait]
+impl<'a, C, I, S, K, V> IOCachedAsync<K, V> for SkytableCache<'a, C, K, V>
+where
+    S: AsyncRead + AsyncWrite + Send + Sync + Unpin,
+    I: DerefMut<Target = TcpConnection<S>> + Send + Sync,
+    C: ManageConnection<Connection = I> + Send + Sync,
+    K: Display + Send + Sync,
+    V: Serialize + DeserializeOwned + Send + Sync,
+{
+    type Error = SkytableCacheError;
+
+    #[instrument(skip_all)]
+    async fn cache_get(&self, k: &K) -> std::result::Result<Option<V>, Self::Error> {
+        if self.refresh {
+            self.pool
+                .get()
+                .await
+                .unwrap()
+                .query(&query!(
+                    format!("update {} set ttl += ? where ckey = ?", self.build_model()).as_str(),
```

Comment:
do not increase the ttl with the ttl duration, replace it with current time + ttl